File: frontend/packages/types/src/temporal.ts
// Temporal Dimension Types
// Supports versioning, branching, and futures for multi-dimensional traceability
// Enables navigation across time (past/present/future) and variation (alternatives, forks, experiments)

// =============================================================================
// VERSION BRANCH TYPES
// =============================================================================

/**
 * Branch type classification - similar to git branch naming conventions
 */
export type BranchType =
  | 'main' // Primary production branch
  | 'release' // Release candidate
  | 'feature' // Feature development
  | 'experiment' // A/B testing, "what if" scenarios
  | 'hotfix' // Emergency fixes
  | 'archive'; // Historical reference only

/**
 * Branch lifecycle status
 */
export type BranchStatus =
  | 'active' // Being actively worked on
  | 'review' // Under review for merge
  | 'merged' // Successfully merged into target
  | 'abandoned' // Discarded, not completed
  | 'archived'; // Kept for historical reference

/**
 * A version branch represents a parallel line of development
 * Similar to git branches but for traceability items
 */
export interface VersionBranch {
  id: string;
  projectId: string;

  // Identity
  name: string; // "main", "feature/new-payment", "experiment/ai-routing"
  slug: string; // URL-safe: "feature-new-payment"
  description?: string | undefined;

  // Branch type classification
  branchType: BranchType;

  // Tree structure
  parentBranchId?: string | undefined; // What branch this was created from
  baseVersionId?: string | undefined; // Which version it forked from

  // State
  status: BranchStatus;
  isDefault: boolean; // Is this the "main" branch
  isProtected: boolean; // Prevent direct edits (require merge)

  // Merge tracking
  mergedAt?: string | undefined; // When merged back
  mergedInto?: string | undefined; // Which branch it merged into
  mergedBy?: string | undefined; // Who performed the merge

  // Statistics
  versionCount: number; // Number of versions in this branch
  itemCount: number; // Current number of items

  // Metadata
  metadata?: Record<string, unknown> | undefined;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

// =============================================================================
// VERSION TYPES
// =============================================================================

/**
 * Version status - lifecycle of a version/commit
 */
export type VersionStatus =
  | 'draft' // Work in progress, not finalized
  | 'pending_review' // Submitted for review/approval
  | 'approved' // Approved and ready for use
  | 'rejected' // Failed review, needs changes
  | 'superseded'; // Replaced by a newer version

/**
 * A version represents a point-in-time snapshot of items
 * Similar to a git commit
 */
export interface Version {
  id: string;
  branchId: string;
  projectId: string;

  // Sequence
  versionNumber: number; // Sequential within branch: 1, 2, 3...

  // Tree structure
  parentVersionId?: string | undefined; // Previous version (linear history)
  childVersionIds?: string[] | undefined; // Branches from this point

  // Snapshot reference
  snapshotId: string; // Reference to full state snapshot
  changesetId?: string | undefined; // Delta from parent (for efficiency)

  // Identity
  tag?: string | undefined; // Human-readable tag: "v1.0.0", "MVP", "Q1-Release"
  message: string; // Commit-style message describing changes

  // Statistics
  itemCount: number; // Number of items in this version
  changeCount: number; // Number of changes from parent

  // Quality gates
  status: VersionStatus;
  approvedBy?: string | undefined;
  approvedAt?: string | undefined;
  rejectionReason?: string | undefined;

  // Metadata
  metadata?: Record<string, unknown> | undefined;
  createdBy: string;
  createdAt: string;
}

/**
 * Version changeset - delta between two versions
 */
export interface VersionChangeset {
  id: string;
  versionId: string;
  parentVersionId: string;

  // Changes
  addedItemIds: string[]; // Items added in this version
  removedItemIds: string[]; // Items removed in this version
  modifiedItemIds: string[]; // Items modified in this version

  // Statistics
  addedCount: number;
  removedCount: number;
  modifiedCount: number;
  totalChanges: number;

  // Detailed changes (optional, for diff view)
  modifications?: ItemModification[] | undefined;

  createdAt: string;
}

/**
 * Detailed modification record for an item
 */
export interface ItemModification {
  itemId: string;
  itemTitle: string;

  // Field-level changes
  fieldChanges: FieldChange[];

  // Summary
  changeType: 'content' | 'metadata' | 'status' | 'relationships' | 'mixed';
}

/**
 * Single field change
 */
export interface FieldChange {
  field: string;
  oldValue: unknown;
  newValue: unknown;
  changeType: 'added' | 'removed' | 'modified';
}

// =============================================================================
// ITEM VERSION MAPPING
// =============================================================================

/**
 * Maps an item to a specific version
 * Stores the item's state at that point in time
 */
export interface ItemVersion {
  id: string;
  itemId: string;
  versionId: string;
  branchId: string;
  projectId: string;

  // Item state at this version (snapshot)
  state: ItemSnapshot;

  // Lifecycle
  lifecycle: ItemLifecycle;

  // Version tracking
  introducedInVersionId: string; // First appeared in this version
  lastModifiedInVersionId: string; // Last changed in this version

  // Metadata
  createdAt: string;
}

/**
 * Snapshot of an item's state at a point in time
 */
export interface ItemSnapshot {
  // Core fields
  title: string;
  description?: string | undefined;
  type: string;
  status: string;
  priority?: string | undefined;

  // Hierarchy
  parentId?: string | undefined;

  // Dimensions
  dimensions?: Record<string, unknown> | undefined;

  // References
  canonicalId?: string | undefined;
  codeRef?: Record<string, unknown> | undefined;
  docRef?: Record<string, unknown> | undefined;

  // Full metadata
  metadata?: Record<string, unknown> | undefined;
}

// =============================================================================
// ITEM LIFECYCLE
// =============================================================================

/**
 * Item lifecycle - progression through development stages
 */
export type ItemLifecycle =
  | 'idea' // Just an idea, no commitment
  | 'proposed' // Formally proposed for consideration
  | 'planned' // Scheduled for implementation
  | 'in_progress' // Being actively worked on
  | 'implemented' // Code/content complete
  | 'verified' // Tested/validated
  | 'released' // In production/published
  | 'deprecated' // Being phased out
  | 'retired'; // No longer active

/**
 * Lifecycle transition rules
 */
export const LIFECYCLE_TRANSITIONS: Record<ItemLifecycle, ItemLifecycle[]> = {
  deprecated: ['retired', 'released'],
  idea: ['proposed', 'retired'],
  implemented: ['verified', 'in_progress', 'retired'],
  in_progress: ['implemented', 'planned', 'retired'],
  planned: ['in_progress', 'retired'],
  proposed: ['planned', 'retired'],
  released: ['deprecated'],
  retired: [],
  verified: ['released', 'implemented', 'retired'],
};

// =============================================================================
// ITEM ALTERNATIVES (VARIATIONS/EXPERIMENTS)
// =============================================================================

/**
 * Relationship between alternative items
 */
export type AlternativeRelationship =
  | 'alternative_to' // Alternative approach to same problem
  | 'supersedes' // Newer version replacing older
  | 'experiment_of'; // Experimental variation

/**
 * Tracks alternative implementations or approaches
 */
export interface ItemAlternative {
  id: string;
  projectId: string;

  // The items being compared
  baseItemId: string; // Original/base item
  alternativeItemId: string; // Alternative item

  // Relationship
  relationship: AlternativeRelationship;
  description?: string | undefined;

  // Selection tracking
  isChosen: boolean; // Was this alternative selected
  chosenAt?: string | undefined;
  chosenBy?: string | undefined;
  chosenReason?: string | undefined;

  // Comparison metrics
  metrics?: {
    effort?: number | undefined; // Estimated effort (story points, hours)
    risk?: 'low' | 'medium' | 'high' | undefined;
    coverage?: number | undefined; // How well it addresses the need
    complexity?: number | undefined;
  };

  // Metadata
  createdAt: string;
  updatedAt: string;
}

// =============================================================================
// TEMPORAL ITEM EXTENSION
// =============================================================================

/**
 * Temporal metadata that can be added to any Item
 */
export interface TemporalMetadata {
  // Version tracking
  versionId: string; // Which version this item exists in
  branchId: string; // Which branch
  introducedInVersion: string; // First appeared
  lastModifiedVersion: string; // Last change

  // Lifecycle
  lifecycle: ItemLifecycle;
  lifecycleChangedAt?: string | undefined;
  lifecycleChangedBy?: string | undefined;

  // Alternatives
  alternatives?: {
    itemId: string;
    relationship: AlternativeRelationship;
    isChosen?: boolean | undefined;
    chosenAt?: string | undefined;
  }[];

  // Future planning
  plannedFor?: {
    targetVersion?: string | undefined; // Which version it's planned for
    targetDate?: string | undefined; // Target completion date
    probability?: number | undefined; // 0-1 likelihood of inclusion
    dependencies?: string[] | undefined; // Item IDs that must complete first
  };
}

// =============================================================================
// VERSION COMPARISON & DIFF
// =============================================================================

/**
 * Type of change in a diff
 */
export type DiffChangeType = 'added' | 'removed' | 'modified';

/**
 * Significance level of a change
 */
export type DiffSignificance = 'minor' | 'moderate' | 'major' | 'breaking';

/**
 * Single field change
 */
export interface FieldDiffChange {
  field: string;
  oldValue: unknown;
  newValue: unknown;
  changeType: 'added' | 'removed' | 'modified';
}

/**
 * Item in a diff result
 */
export interface DiffItem {
  itemId: string;
  type: string;
  title: string;
  changeType: DiffChangeType;

  // For modified items - field-level changes
  fieldChanges?: FieldDiffChange[] | undefined;

  // Severity/impact
  significance: DiffSignificance;
}

/**
 * Statistics about a diff
 */
export interface DiffStatistics {
  totalChanges: number;
  addedCount: number;
  removedCount: number;
  modifiedCount: number;
  unchangedCount: number;
}

/**
 * Result of comparing two versions
 */
export interface VersionDiff {
  versionA: string; // Source version ID
  versionB: string; // Target version ID
  versionANumber: number;
  versionBNumber: number;

  // Changes
  added: DiffItem[]; // New in B
  removed: DiffItem[]; // Gone in B
  modified: DiffItem[]; // Changed between A and B
  unchanged: number; // Count of unchanged items

  // Statistics
  stats: DiffStatistics;

  // Computed at
  computedAt: string;
}

/**
 * Diff view configuration
 */
export interface DiffViewConfig {
  showUnchanged: boolean;
  expandModified: boolean;
  filterByType?: string | undefined;
  sortBy?: 'title' | 'changeType' | 'significance' | undefined;
}

/**
 * Options for diff export
 */
export interface DiffExportOptions {
  format: 'json' | 'csv' | 'markdown' | 'html';
  includeUnchanged: boolean;
  includeFieldChanges: boolean;
}

/**
 * Diff export result
 */
export interface DiffExportResult {
  filename: string;
  mimeType: string;
  content: string | Blob;
}

/**
 * UI state for diff viewer
 */
export interface DiffViewerState {
  expandedItems: Set<string>;
  selectedItems: Set<string>;
  activeTab: 'added' | 'removed' | 'modified';
  searchQuery: string;
}

// =============================================================================
// MERGE OPERATIONS
// =============================================================================

/**
 * Merge request - proposal to merge one branch into another
 */
export interface MergeRequest {
  id: string;
  projectId: string;

  // Branches
  sourceBranchId: string;
  targetBranchId: string;

  // Versions
  sourceVersionId: string; // Head of source branch
  baseVersionId: string; // Common ancestor

  // Status
  status: 'open' | 'approved' | 'merged' | 'closed' | 'conflict';

  // Details
  title: string;
  description?: string | undefined;

  // Diff
  diff?: VersionDiff | undefined;

  // Conflicts (if any)
  conflicts?: MergeConflict[] | undefined;

  // Review
  reviewers?: string[] | undefined;
  approvedBy?: string[] | undefined;

  // Timestamps
  createdBy: string;
  createdAt: string;
  mergedAt?: string | undefined;
  mergedBy?: string | undefined;
  closedAt?: string | undefined;
}

/**
 * Conflict during merge
 */
export interface MergeConflict {
  itemId: string;
  itemTitle: string;

  // Conflicting values
  sourceValue: unknown;
  targetValue: unknown;
  baseValue?: unknown | undefined; // Value in common ancestor

  // Resolution
  resolution?: 'source' | 'target' | 'manual' | 'both' | undefined;
  resolvedValue?: unknown | undefined;
  resolvedBy?: string | undefined;
  resolvedAt?: string | undefined;
}

// =============================================================================
// TEMPORAL NAVIGATION
// =============================================================================

/**
 * Temporal navigation mode
 */
export type TemporalViewMode =
  | 'timeline' // Horizontal timeline with version markers
  | 'branches' // Tree visualization of branches
  | 'comparison' // Side-by-side diff of two versions
  | 'progress'; // Milestone/sprint tracking dashboard

/**
 * Temporal filter for querying items at a point in time
 */
export interface TemporalFilter {
  // Time-based
  atVersion?: string | undefined; // Specific version ID
  atDate?: string | undefined; // Point in time
  afterVersion?: string | undefined; // After this version
  beforeVersion?: string | undefined; // Before this version

  // Branch-based
  branchId?: string | undefined; // Specific branch
  includeMerged?: boolean | undefined; // Include merged branches

  // Lifecycle-based
  lifecycle?: ItemLifecycle[] | undefined; // Filter by lifecycle stages

  // Status-based
  versionStatus?: VersionStatus[] | undefined; // Filter by version status
}

// =============================================================================
// UTILITY FUNCTIONS
// =============================================================================

/**
 * Check if a lifecycle transition is valid
 *
 * @param {ItemLifecycle} from - Current lifecycle stage.
 * @param {ItemLifecycle} to - Target lifecycle stage.
 * @returns {boolean} True when transition from `from` to `to` is allowed.
 */
export function isValidLifecycleTransition(from: ItemLifecycle, to: ItemLifecycle): boolean {
  return LIFECYCLE_TRANSITIONS[from]?.includes(to) ?? false;
}

/**
 * Get display color for branch type
 *
 * @param {BranchType} branchType - Branch type to map.
 * @returns {string} Hex color for branch type.
 */
export function getBranchTypeColor(branchType: BranchType): string {
  const colors: Record<BranchType, string> = {
    main: '#22c55e', // Green
    release: '#3b82f6', // Blue
    feature: '#a855f7', // Purple
    experiment: '#f59e0b', // Amber
    hotfix: '#ef4444', // Red
    archive: '#6b7280', // Gray
  };
  return colors[branchType];
}

/**
 * Get display color for lifecycle stage
 *
 * @param {ItemLifecycle} lifecycle - Lifecycle stage to map.
 * @returns {string} Hex color for lifecycle stage.
 */
export function getLifecycleColor(lifecycle: ItemLifecycle): string {
  const colors: Record<ItemLifecycle, string> = {
    idea: '#94a3b8', // Slate
    proposed: '#a78bfa', // Violet
    planned: '#60a5fa', // Blue
    in_progress: '#fbbf24', // Amber
    implemented: '#34d399', // Emerald
    verified: '#22c55e', // Green
    released: '#10b981', // Emerald dark
    deprecated: '#f97316', // Orange
    retired: '#6b7280', // Gray
  };
  return colors[lifecycle];
}

/**
 * Get display icon for branch type
 *
 * @param {BranchType} branchType - Branch type to map.
 * @returns {string} Icon name for branch type.
 */
export function getBranchTypeIcon(branchType: BranchType): string {
  const icons: Record<BranchType, string> = {
    archive: 'Archive',
    experiment: 'Flask',
    feature: 'Sparkles',
    hotfix: 'Zap',
    main: 'GitBranch',
    release: 'Tag',
  };
  return icons[branchType];
}

File: backend/internal/models/temporal.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// VersionBranch represents a version control branch for managing parallel development
type VersionBranch struct {
	ID          string  `gorm:"column:id;primaryKey;type:uuid" json:"id"`
	ProjectID   string  `gorm:"column:project_id;type:uuid;not null" json:"project_id"`
	Name        string  `gorm:"column:name;size:255;not null" json:"name"`
	Slug        string  `gorm:"column:slug;size:255;not null" json:"slug"`
	Description *string `gorm:"column:description;type:text" json:"description,omitempty"`
	// BranchType: main, release, feature, experiment, hotfix, archive
	BranchType     string  `gorm:"column:branch_type;size:50;not null;default:'feature'" json:"branch_type"`
	ParentBranchID *string `gorm:"column:parent_branch_id;type:uuid" json:"parent_branch_id,omitempty"`
	BaseVersionID  *string `gorm:"column:base_version_id;type:uuid" json:"base_version_id,omitempty"`
	// Status: active, review, merged, abandoned, archived
	Status       string         `gorm:"column:status;size:50;not null;default:'active'" json:"status"`
	IsDefault    bool           `gorm:"column:is_default;not null;default:false" json:"is_default"`
	IsProtected  bool           `gorm:"column:is_protected;not null;default:false" json:"is_protected"`
	MergedAt     *time.Time     `gorm:"column:merged_at" json:"merged_at,omitempty"`
	MergedInto   *string        `gorm:"column:merged_into;type:uuid" json:"merged_into,omitempty"`
	MergedBy     *string        `gorm:"column:merged_by;size:255" json:"merged_by,omitempty"`
	VersionCount int            `gorm:"column:version_count;not null;default:0" json:"version_count"`
	ItemCount    int            `gorm:"column:item_count;not null;default:0" json:"item_count"`
	Metadata     datatypes.JSON `gorm:"column:metadata;type:jsonb;default:'{}'" json:"metadata"`
	CreatedBy    string         `gorm:"column:created_by;size:255;not null" json:"created_by"`
	CreatedAt    time.Time      `gorm:"column:created_at;not null;autoCreateTime" json:"created_at"`
	UpdatedAt    time.Time      `gorm:"column:updated_at;not null;autoUpdateTime" json:"updated_at"`

	// Relationships
	Project      *Project       `gorm:"foreignKey:ProjectID;references:ID" json:"project,omitempty"`
	ParentBranch *VersionBranch `gorm:"foreignKey:ParentBranchID;references:ID" json:"parent_branch,omitempty"`
}

// TableName overrides GORM table name
func (VersionBranch) TableName() string {
	return "version_branches"
}

// BeforeCreate generates UUID for ID if not set
func (vb *VersionBranch) BeforeCreate(_ *gorm.DB) error {
	if vb.ID == "" {
		vb.ID = uuid.New().String()
	}
	return nil
}

// Version represents a snapshot/commit in a version branch
type Version struct {
	ID              string  `gorm:"column:id;primaryKey;type:uuid" json:"id"`
	BranchID        string  `gorm:"column:branch_id;type:uuid;not null" json:"branch_id"`
	ProjectID       string  `gorm:"column:project_id;type:uuid;not null" json:"project_id"`
	VersionNumber   int     `gorm:"column:version_number;not null" json:"version_number"`
	ParentVersionID *string `gorm:"column:parent_version_id;type:uuid" json:"parent_version_id,omitempty"`
	SnapshotID      *string `gorm:"column:snapshot_id;type:uuid" json:"snapshot_id,omitempty"`
	ChangesetID     *string `gorm:"column:changeset_id;type:uuid" json:"changeset_id,omitempty"`
	Tag             *string `gorm:"column:tag;size:255" json:"tag,omitempty"`
	Message         string  `gorm:"column:message;type:text;not null" json:"message"`
	ItemCount       int     `gorm:"column:item_count;not null;default:0" json:"item_count"`
	ChangeCount     int     `gorm:"column:change_count;not null;default:0" json:"change_count"`
	// Status: draft, pending_review, approved, rejected, superseded
	Status          string         `gorm:"column:status;size:50;not null;default:'draft'" json:"status"`
	ApprovedBy      *string        `gorm:"column:approved_by;size:255" json:"approved_by,omitempty"`
	ApprovedAt      *time.Time     `gorm:"column:approved_at" json:"approved_at,omitempty"`
	RejectionReason *string        `gorm:"column:rejection_reason;type:text" json:"rejection_reason,omitempty"`
	Metadata        datatypes.JSON `gorm:"column:metadata;type:jsonb;default:'{}'" json:"metadata"`
	CreatedBy       string         `gorm:"column:created_by;size:255;not null" json:"created_by"`
	CreatedAt       time.Time      `gorm:"column:created_at;not null;autoCreateTime" json:"created_at"`

	// Relationships
	Branch        *VersionBranch `gorm:"foreignKey:BranchID;references:ID" json:"branch,omitempty"`
	Project       *Project       `gorm:"foreignKey:ProjectID;references:ID" json:"project,omitempty"`
	ParentVersion *Version       `gorm:"foreignKey:ParentVersionID;references:ID" json:"parent_version,omitempty"`
}

// TableName overrides GORM table name
func (Version) TableName() string {
	return "versions"
}

// BeforeCreate generates UUID for ID if not set
func (v *Version) BeforeCreate(_ *gorm.DB) error {
	if v.ID == "" {
		v.ID = uuid.New().String()
	}
	return nil
}

// ItemVersion represents a snapshot of an item at a specific version
type ItemVersion struct {
	ID                      string         `gorm:"column:id;primaryKey;type:uuid" json:"id"`
	ItemID                  string         `gorm:"column:item_id;type:uuid;not null" json:"item_id"`
	VersionID               string         `gorm:"column:version_id;type:uuid;not null" json:"version_id"`
	BranchID                string         `gorm:"column:branch_id;type:uuid;not null" json:"branch_id"`
	ProjectID               string         `gorm:"column:project_id;type:uuid;not null" json:"project_id"`
	State                   datatypes.JSON `gorm:"column:state;type:jsonb;not null" json:"state"`
	Lifecycle               *string        `gorm:"column:lifecycle;size:50" json:"lifecycle,omitempty"`
	IntroducedInVersionID   *string        `gorm:"column:introduced_in_version_id;type:uuid" json:"introduced_in_version_id,omitempty"`
	LastModifiedInVersionID *string        `gorm:"type:uuid" json:"last_modified_in_version_id,omitempty"`
	CreatedAt               time.Time      `gorm:"column:created_at;not null;autoCreateTime" json:"created_at"`

	// Relationships
	Item    *Item          `gorm:"foreignKey:ItemID;references:ID" json:"item,omitempty"`
	Version *Version       `gorm:"foreignKey:VersionID;references:ID" json:"version,omitempty"`
	Branch  *VersionBranch `gorm:"foreignKey:BranchID;references:ID" json:"branch,omitempty"`
	Project *Project       `gorm:"foreignKey:ProjectID;references:ID" json:"project,omitempty"`
}

// TableName overrides GORM table name
func (ItemVersion) TableName() string {
	return "item_versions"
}

// BeforeCreate generates UUID for ID if not set
func (iv *ItemVersion) BeforeCreate(_ *gorm.DB) error {
	if iv.ID == "" {
		iv.ID = uuid.New().String()
	}
	return nil
}

// ItemAlternative represents alternative implementations or design choices for an item
type ItemAlternative struct {
	ID                string `gorm:"column:id;primaryKey;type:uuid" json:"id"`
	ProjectID         string `gorm:"column:project_id;type:uuid;not null" json:"project_id"`
	BaseItemID        string `gorm:"column:base_item_id;type:uuid;not null" json:"base_item_id"`
	AlternativeItemID string `gorm:"column:alternative_item_id;type:uuid;not null" json:"alternative_item_id"`
	// Relationship: alternative_to, supersedes, experiment_of
	Relationship string         `gorm:"column:relationship;size:100;not null" json:"relationship"`
	Description  *string        `gorm:"column:description;type:text" json:"description,omitempty"`
	IsChosen     bool           `gorm:"column:is_chosen;not null;default:false" json:"is_chosen"`
	ChosenAt     *time.Time     `gorm:"column:chosen_at" json:"chosen_at,omitempty"`
	ChosenBy     *string        `gorm:"column:chosen_by;size:255" json:"chosen_by,omitempty"`
	ChosenReason *string        `gorm:"column:chosen_reason;type:text" json:"chosen_reason,omitempty"`
	Metrics      datatypes.JSON `gorm:"column:metrics;type:jsonb;default:'{}'" json:"metrics"`
	CreatedAt    time.Time      `gorm:"column:created_at;not null;autoCreateTime" json:"created_at"`
	UpdatedAt    time.Time      `gorm:"column:updated_at;not null;autoUpdateTime" json:"updated_at"`

	// Relationships
	Project         *Project `gorm:"foreignKey:ProjectID;references:ID" json:"project,omitempty"`
	BaseItem        *Item    `gorm:"foreignKey:BaseItemID;references:ID" json:"base_item,omitempty"`
	AlternativeItem *Item    `gorm:"foreignKey:AlternativeItemID;references:ID" json:"alternative_item,omitempty"`
}

// TableName overrides GORM table name
func (ItemAlternative) TableName() string {
	return "item_alternatives"
}

// BeforeCreate generates UUID for ID if not set
func (ia *ItemAlternative) BeforeCreate(_ *gorm.DB) error {
	if ia.ID == "" {
		ia.ID = uuid.New().String()
	}
	return nil
}

// MergeRequest represents a request to merge one branch into another
type MergeRequest struct {
	ID              string  `gorm:"column:id;primaryKey;type:uuid" json:"id"`
	ProjectID       string  `gorm:"column:project_id;type:uuid;not null" json:"project_id"`
	SourceBranchID  string  `gorm:"column:source_branch_id;type:uuid;not null" json:"source_branch_id"`
	TargetBranchID  string  `gorm:"column:target_branch_id;type:uuid;not null" json:"target_branch_id"`
	SourceVersionID string  `gorm:"column:source_version_id;type:uuid;not null" json:"source_version_id"`
	BaseVersionID   *string `gorm:"column:base_version_id;type:uuid" json:"base_version_id,omitempty"`
	// Status: open, approved, merged, closed, conflict
	Status      string         `gorm:"column:status;size:50;not null;default:'open'" json:"status"`
	Title       string         `gorm:"column:title;size:255;not null" json:"title"`
	Description *string        `gorm:"column:description;type:text" json:"description,omitempty"`
	Diff        datatypes.JSON `gorm:"column:diff;type:jsonb;default:'{}'" json:"diff"`
	Conflicts   datatypes.JSON `gorm:"column:conflicts;type:jsonb;default:'[]'" json:"conflicts"`
	Reviewers   datatypes.JSON `gorm:"column:reviewers;type:jsonb;default:'[]'" json:"reviewers"`
	ApprovedBy  datatypes.JSON `gorm:"column:approved_by;type:jsonb;default:'[]'" json:"approved_by"`
	CreatedBy   string         `gorm:"column:created_by;size:255;not null" json:"created_by"`
	CreatedAt   time.Time      `gorm:"column:created_at;not null;autoCreateTime" json:"created_at"`
	MergedAt    *time.Time     `gorm:"column:merged_at" json:"merged_at,omitempty"`
	MergedBy    *string        `gorm:"column:merged_by;size:255" json:"merged_by,omitempty"`
	ClosedAt    *time.Time     `gorm:"column:closed_at" json:"closed_at,omitempty"`
	UpdatedAt   time.Time      `gorm:"column:updated_at;not null;autoUpdateTime" json:"updated_at"`

	// Relationships
	Project       *Project       `gorm:"foreignKey:ProjectID;references:ID" json:"project,omitempty"`
	SourceBranch  *VersionBranch `gorm:"foreignKey:SourceBranchID;references:ID" json:"source_branch,omitempty"`
	TargetBranch  *VersionBranch `gorm:"foreignKey:TargetBranchID;references:ID" json:"target_branch,omitempty"`
	SourceVersion *Version       `gorm:"foreignKey:SourceVersionID;references:ID" json:"source_version,omitempty"`
	BaseVersion   *Version       `gorm:"foreignKey:BaseVersionID;references:ID" json:"base_version,omitempty"`
}

// TableName overrides GORM table name
func (MergeRequest) TableName() string {
	return "merge_requests"
}

// BeforeCreate generates UUID for ID if not set
func (mr *MergeRequest) BeforeCreate(_ *gorm.DB) error {
	if mr.ID == "" {
		mr.ID = uuid.New().String()
	}
	return nil
}
